packages/nbdime/src/common/mergeview.ts

```diff
@@ -143,563 +341,404 @@ function createNbdimeMergeView(
-      // Merge editor does not set a marker currently, so don't clear for it:
-      if (valueIn(classes.gutter, [leftClasses.gutter, rightClasses.gutter])) {
-        editor.setGutterMarker(mark, GUTTER_PICKER_CLASS, null);
-      } else {
-        editor.setGutterMarker(mark, GUTTER_CONFLICT_CLASS, null);
-      }
-      let line = editor.lineInfo(mark);
-      if (!line.bgClass || line.bgClass.length === 0) {
-        arr.splice(i, 1);
+  private buildCharacterHighlighting(
+    editor: EditorView,
+    diffRanges: DiffRangePos[],
+    markType: DIFF_OP
+  ) {
+    let effects: StateEffect<unknown>[] = [];
+    let sources: ChunkSource[] = [];
+    if (markType === DIFF_OP.DIFF_INSERT || markType === DIFF_OP.DIFF_DELETE) {
+      let highlightType: string = markType === DIFF_OP.DIFF_DELETE ? 'deleted' : 'inserted';
+
+      for (let r of diffRanges) {
+        if (r.source !== undefined) {
+          sources.push(r.source);
+        }
+        let decorationKey = this.getDecorationKey(sources);
+        let startingOffset = posToOffset(editor.state.doc, {
+          line: r.from.line,
+          column: r.from.column
+        });
+        let endingOffset = posToOffset(editor.state.doc, {
+          line: r.to.line,
+          column: r.to.column
+        });
+        effects.push(addHighlightEffect.of({
+          from: startingOffset,
+          to: endingOffset,
+          highlightType: highlightType,
+          decorationKey: decorationKey
+        }));
       }
     }
+    return effects;
   }
-}
 
-function highlightChars(editor: CodeMirror.Editor, ranges: DiffRangePos[],
-                        markers: Marker[], cls: string) {
-  let doc = editor.getDoc();
-  let origCls: string | null = null;
-  if (valueIn(cls, [mergeClassPrefix.del, mergeClassPrefix.insert])) {
-    origCls = cls;
-  }
-  for (let r of ranges) {
-    if (origCls !== null) {
-      cls = origCls + (r.source ? '-' + r.source.action : '');
+  private clearCharacterHighlighting(
+    editor: EditorView,
+    diffRanges: DiffRangePos[],
+    markType: DIFF_OP
+  ) {
+    let effects: StateEffect<unknown>[] = [];
+    let sources: ChunkSource[] = [];
+    if (markType === DIFF_OP.DIFF_INSERT || markType === DIFF_OP.DIFF_DELETE) {
+      let highlightType: string = markType === DIFF_OP.DIFF_DELETE ? 'deleted' : 'inserted';
+      for (let r of diffRanges) {
+        if (r.source !== undefined) {
+          sources.push(r.source);
+        }
+        let decorationKey = this.getDecorationKey(sources);
+        effects.push(removeHighlightEffect.of({highlightType: highlightType, decorationKey: decorationKey}));
+      }
     }
-    markers.push(doc.markText(r.from, r.to, {className: cls}));
+    return effects;
   }
-}
-
 
-// Updating the gap between editor and original
+  protected updateHighlighting(
+    editor: EditorView,
+    diffRanges: DiffRangePos[],
+    chunkArray: Chunk[],
+    type: DIFF_OP
+  ) {
+    let self = this;
```

Comment:
No need to do that

Reply:
Thanks. The definition and use of `self` has been removed.

Reply:
Thanks. The definition and use of `self` has been removed.